Move the technical docs to serverless-manager (#166)
kyma-project/serverless-manager: docs/user/07-70-function-specification.md

# Function's specification

> Serverless in Kyma allows you to create Functions in both Node.js and Python. Although the Function's interface is unified, its specification differs depending on the runtime used to run the Function.

Reviewer: ```suggestion
With the Serverless module, you can create Functions in both Node.js and Python. Although the Function's interface is unified, its specification differs depending on the runtime used to run the Function.
```

```json
...
{ "function-name": "main",
  "timeout": 180,
  "runtime": "nodejs18",
  "memory-limit": 200Mi }
```

See the detailed descriptions of these fields:

| Field | Description                                                                               |
|-------|-------------------------------------------------------------------------------------------|
| **function-name** | Name of the invoked Function                                                              |
| **timeout** | Time, in seconds, after which the system cancels the request to invoke the Function       |
| **runtime** | Environment used to run the Function. You can use `nodejs16` - deprecated, `nodejs18`, or `python39`. |
| **memory-limit** | Maximum amount of memory assigned to run a Function                                       |

## HTTP requests

You can use the **event.extensions.request** object to access properties and methods of a given request that vary depending on the runtime. For more information, read the API documentation for [Node.js Express](http://expressjs.com/en/api.html#req) and [Python](https://bottlepy.org/docs/dev/api.html#the-request-object).

## Custom HTTP responses

> 

Reviewer: ```suggestion

By default, a failing Function simply throws an error to tell the Event Service to reinject the event at a later point. Such an HTTP-based Function returns the HTTP status code `500`.  If you manage to invoke a Function successfully, the system returns the default HTTP status code `200`.

Apart from these two default codes, you can define custom responses. Learn how to do that in Node.js and Python:

```

<div tabs name="custom-http-response" group="function-specification">
<details>
<summary label="Node.js">
Node.js
</summary>

> By default, a failing Function simply throws an error to tell the Event Service to reinject the event at a later point. Such an HTTP-based Function returns the HTTP status code `500`. On the contrary, if you manage to invoke a Function successfully, the system returns the default HTTP status code `200`.

Reviewer: As this paragraph is identical for node.js and Python, you can extract it from the divs and use it as introductory paragraph above them.

```suggestion
```

> Apart from these two default codes, you can define custom responses in all Node.js runtimes using the **event.extensions.response** object.

Reviewer: ```suggestion
To define custom responses in all Node.js runtimes, use the **event.extensions.response** object.
```

This object is created by the Express framework and can be customized. For more information, read [Node.js API documentation](https://nodejs.org/docs/latest-v12.x/api/http.html#http_class_http_serverresponse).

This example shows how to set such a custom response in Node.js for the HTTP status code `400`:

```js
module.exports = {    
    main: function (event, context) {
        if (event.extensions.request.query.id === undefined) {
            res = event.extensions.response;
            res.status(400)
            return
        }
        return "42"
    }
}
```

</details>
<details>
<summary label="Python">
Python
</summary>

> By default, a failing Function simply throws an error to tell the Event Service to reinject the event at a later point. Such an HTTP-based Function returns the HTTP status code `500`. In case you manage to invoke a Function successfully, the system returns the default HTTP status code `200`.

Reviewer: ```suggestion
```

added in line 175

> Apart from these two default codes, you can define custom responses in all Python runtimes using the **HTTPResponse** object available in Bottle.

Reviewer: ```suggestion
To define custom responses in all Python runtimes, use the **HTTPResponse** object available in Bottle.
```

> This object will need to be instantiated and can be customized. For more information, read [Bottle API documentation](https://bottlepy.org/docs/dev/api.html#the-response-object).

Reviewer: ```suggestion
This object must be instantiated and can be customized. For more information, read [Bottle API documentation](https://bottlepy.org/docs/dev/api.html#the-response-object).
```

The following example shows how to set such a custom response in Python for the HTTP status code `400`:

```python
from bottle import HTTPResponse

SUPPORTED_CONTENT_TYPES = ['application/json']

def main(event, context):
    request = event['extensions']['request']

    response_content_type = 'application/json'
    headers = {
        'Content-Type': response_content_type
    }

    status = 202
    response_payload = {'success': 'Message accepted.'}

    if request.headers.get('Content-Type') not in SUPPORTED_CONTENT_TYPES:
        status = 400
        response_payload = json.dumps({'error': 'Invalid Content-Type.'})

    return HTTPResponse(body=response_payload, status=status, headers=headers)
```

</details>
</div>

## Override runtime image

You can use a custom runtime image to override the existing one. Your image must meet all the following requirements:

- Expose the workload endpoint on the right port
- Provide liveness and readiness check endpoints at `/healthz`
- Fetch sources from the path under the `KUBELESS_INSTALL_VOLUME` environment
- Security support. Kyma runtimes are secure by default. You only need to protect your images.

> > Note: For better understanding, you can look at [main dockerfiles](https://github.com/kyma-project/kyma/blob/main/resources/serverless/templates/runtimes.yaml) which are responsible for building the final image based on the `base_image` argument which you as a user can override and what we are doing in [this tutorial](/docs/user/03-110-override-runtime-image.md).

Reviewer: ```suggestion
> **NOTE:** For better understanding, you can look at the [main Docker files](https://github.com/kyma-project/kyma/blob/main/resources/serverless/templates/runtimes.yaml), which are responsible for building the final image based on the `base_image` argument, which you, as a user, can override and what we are doing in [this tutorial](/docs/user/03-110-override-runtime-image.md).
```

I think we should simplify this sentence and split it in two or three shorter ones.

> Every Function's Pods container have the same system environments which helps you configure the Functions server. For more information, read the [Environment variables](/docs/user/05-20-env-variables.md) page.

Reviewer: ```suggestion
Every Function's Pods container has the same system environments, which helps you configure the Functions server. For more information, read the [Environment variables](/docs/user/05-20-env-variables.md) page.
```

Apart from these two default codes, you can define custom responses in all Python runtimes using the **HTTPResponse** object available in Bottle.

This object will need to be instantiated and can be customized. For more information, read [Bottle API documentation](https://bottlepy.org/docs/dev/api.html#the-response-object).

The following example shows how to set such a custom response in Python for the HTTP status code `400`:

```python
from bottle import HTTPResponse

SUPPORTED_CONTENT_TYPES = ['application/json']

def main(event, context):
    request = event['extensions']['request']

    response_content_type = 'application/json'
    headers = {
        'Content-Type': response_content_type
    }

    status = 202
    response_payload = {'success': 'Message accepted.'}

    if request.headers.get('Content-Type') not in SUPPORTED_CONTENT_TYPES:
        status = 400
        response_payload = json.dumps({'error': 'Invalid Content-Type.'})

    return HTTPResponse(body=response_payload, status=status, headers=headers)
```

</details>
</div>

> ## /metrics endpoint  
> 
> You can use the `/metrics` endpoint to return the Function metrics. All the information is gathered using Prometheus and can be displayed using the Grafana dashboard (see [Kyma observability](https://kyma-project.io/docs/kyma/latest/02-get-started/05-observability/) for more information on how to use Grafana dashboard in Kyma). As this endpoint is provided by Kubeless, it cannot be customized.  
> For more information, see [Kubeless monitoring](https://github.com/vmware-archive/kubeless/blob/master/docs/monitoring.md) and [Kubeless runtime variants](https://github.com/vmware-archive/kubeless/blob/master/docs/runtimes.md) pages.
> 
> > **NOTE:** Prometheus and Grafana are [deprecated](https://kyma-project.io/blog/2022/12/9/monitoring-deprecation) and are planned to be removed. If you want to install a custom stack, take a look at [Install a custom kube-prometheus-stack in Kyma](https://github.com/kyma-project/examples/tree/main/prometheus).
> 

Reviewer 2: ```suggestion
```

Unfortunately the endpoint is no longer available (since we removed kubeless from nodejs)
